apps/aecore/lib/aecore/oracle/oracle.ex

```diff
@@ -0,0 +1,296 @@
+
+    {:ok, tx} = SignedTx.sign_tx(tx_data, Wallet.get_private_key())
+    Pool.add_transaction(tx)
+  end
+
+  @doc """
+  Creates an oracle response transaction with the query referenced by its
+  transaction hash and the data of the response.
+  """
+  @spec respond(binary(), any(), non_neg_integer()) :: :ok | :error
+  def respond(query_id, response, fee) do
+    payload = %{
+      query_id: query_id,
+      response: response
+    }
+
+    tx_data =
+      DataTx.init(
+        OracleResponseTxData,
+        payload,
+        Wallet.get_public_key(),
+        fee,
+        Chain.lowest_valid_nonce()
+      )
+
+    {:ok, tx} = SignedTx.sign_tx(tx_data, Wallet.get_private_key())
+    Pool.add_transaction(tx)
+  end
+
+  @spec extend(non_neg_integer(), non_neg_integer()) :: :ok | :error
+  def extend(ttl, fee) do
+    payload = %{
+      ttl: ttl
+    }
+
+    tx_data =
+      DataTx.init(
+        OracleExtendTxData,
+        payload,
+        Wallet.get_public_key(),
+        fee,
+        Chain.lowest_valid_nonce()
+      )
+
+    {:ok, tx} = SignedTx.sign_tx(tx_data, Wallet.get_private_key())
+    Pool.add_transaction(tx)
+  end
+
+  @spec data_valid?(map(), map()) :: true | false
+  def data_valid?(format, data) do
+    schema = ExJsonSchema.Schema.resolve(format)
+
+    case ExJsonSchema.Validator.validate(schema, data) do
+      :ok ->
+        true
+
+      {:error, [{message, _}]} ->
+        Logger.error(fn -> message end)
+        false
+    end
+  end
+
+  @spec calculate_absolute_ttl(ttl(), non_neg_integer()) :: non_neg_integer()
+  def calculate_absolute_ttl(%{ttl: ttl, type: type}, block_height_tx_included) do
+    case type do
+      :absolute ->
+        ttl
+
+      :relative ->
+        ttl + block_height_tx_included
+    end
+  end
+
+  @spec calculate_relative_ttl(%{ttl: non_neg_integer(), type: :absolute}, non_neg_integer()) ::
+          non_neg_integer()
+  def calculate_relative_ttl(%{ttl: ttl, type: :absolute}, block_height) do
+    ttl - block_height
+  end
+
+  @spec tx_ttl_is_valid?(oracle_txs_with_ttl(), non_neg_integer()) :: boolean
+  def tx_ttl_is_valid?(tx, block_height) do
+    case tx do
+      %OracleRegistrationTxData{} ->
+        ttl_is_valid?(tx.ttl, block_height)
+
+      %OracleQueryTxData{} ->
+        response_ttl_is_valid =
+          case tx.response_ttl do
+            %{type: :absolute} ->
+              Logger.error("Response TTL has to be relative")
+              false
+
+            %{type: :relative} ->
+              ttl_is_valid?(tx.response_ttl, block_height)
+          end
+
+        query_ttl_is_valid = ttl_is_valid?(tx.query_ttl, block_height)
+
+        response_ttl_is_valid && query_ttl_is_valid
+
+      %OracleExtendTxData{} ->
+        tx.ttl > 0
+
+      _ ->
+        true
+    end
+  end
+
+  @spec ttl_is_valid?(ttl()) :: boolean()
+  def ttl_is_valid?(ttl) do
+    case ttl do
+      %{ttl: ttl, type: :absolute} ->
+        ttl > 0
+
+      %{ttl: ttl, type: :relative} ->
```

Comment:
Can't this case and the first one be done in only one case, being that they have the same outcome? eg `%{ttl: ttl, type: _type} -> `

Reply:
If it's done like that you would be able to input something other than :absolute or :relative for types

Reply:
This is just a pattern match, checking whether the type is `:absolute` or `:relative` should be checked when creating/initializing an Oracle type transaction. At this stage it should be sure that the type is either `:abosolute` or `:relative`, right?

Reply:
This is called in the is_valid? functions